### ISSUE TITLE
FailedPreStopHook event while Stopping container driver-registrar

### ISSUE THREAD

**Issue by Jainbrt** [IBM/ibm-spectrum-scale-csi#924]
## Describe the bug
FailedPreStopHook  event while Stopping container driver-registrar

## How to Reproduce?
Please list the steps to help development teams reproduce the behavior

1. Installed CSI 2.9.0 Dev ( 22022023) build

```
[root@xyz independent]# oc get pods
NAME                                                  READY   STATUS    RESTARTS      AGE
ibm-spectrum-scale-csi-attacher-b788cc49b-c26vs       1/1     Running   0             2d5h
ibm-spectrum-scale-csi-attacher-b788cc49b-h86cp       1/1     Running   0             2d5h
ibm-spectrum-scale-csi-d9nnr                          3/3     Running   0             10m
ibm-spectrum-scale-csi-lh6tn                          3/3     Running   0             10m
ibm-spectrum-scale-csi-operator-69796c5fb-hb84n       1/1     Running   0             22h
ibm-spectrum-scale-csi-p5nrs                          3/3     Running   0             10m
ibm-spectrum-scale-csi-provisioner-6dc64c77df-c2zs6   1/1     Running   3 (21h ago)   2d5h
ibm-spectrum-scale-csi-resizer-6595bc644-bnpw4        1/1     Running   0             2d5h
ibm-spectrum-scale-csi-snapshotter-65f6ffb99b-5w479   1/1     Running   0             2d5h

```
2. Deleted one of the CSI driver plugin pod 

```
[root@xyz independent]# oc delete pod ibm-spectrum-scale-csi-d9nnr
pod "ibm-spectrum-scale-csi-d9nnr" deleted
```

Observed "FailedPreStopHook  event while Stopping container driver-registrar"

```
[root@xyz independent]# oc get events
4s          Normal    Killing                 pod/ibm-spectrum-scale-csi-d9nnr          Stopping container ibm-spectrum-scale-csi
4s          Normal    Killing                 pod/ibm-spectrum-scale-csi-d9nnr          Stopping container liveness-probe
4s          Normal    Killing                 pod/ibm-spectrum-scale-csi-d9nnr          Stopping container driver-registrar
4s          Warning   FailedPreStopHook       pod/ibm-spectrum-scale-csi-d9nnr          Exec lifecycle hook ([/bin/sh -c rm -rf /var/lib/kubelet/plugins/spectrumscale.csi.ibm.com/csi.sock]) for Container "driver-registrar" in Pod "ibm-spectrum-scale-csi-d9nnr_ibm-spectrum-scale-csi(52ff5540-fa01-4a02-a230-cc06efeee53f)" failed - error: rpc error: code = Unknown desc = command error: time="2023-02-23T08:50:28Z" level=error msg="exec failed: unable to start container process: exec: \"/bin/sh\": stat /bin/sh: no such file or directory"...
```

## Expected behavior
We should not see FailedPreStopHook  event while Stopping container driver-registrar

**Comment by amdabhad** [comment]
As preStop hook is moved to driver container, the event like below is not seen:
```
4s          Warning   FailedPreStopHook       pod/ibm-spectrum-scale-csi-d9nnr          Exec lifecycle hook ([/bin/sh -c rm -rf /var/lib/kubelet/plugins/spectrumscale.csi.ibm.com/csi.sock]) for Container "driver-registrar" in Pod "ibm-spectrum-scale-csi-d9nnr_ibm-spectrum-scale-csi(52ff5540-fa01-4a02-a230-cc06efeee53f)" failed - error: rpc error: code = Unknown desc = command error: time="2023-02-23T08:50:28Z" level=error msg="exec failed: unable to start container process: exec: \"/bin/sh\": stat /bin/sh: no such file or directory"
```
and also the expected cleanup happens when driver pods are deleted e.g.:
```
[root@cluster31-worker-1 ~]# ls /var/lib/kubelet/plugins/spectrumscale.csi.ibm.com/csi.sock
ls: cannot access /var/lib/kubelet/plugins/spectrumscale.csi.ibm.com/csi.sock: No such file or directory

```
but there are still some events like below can be seen from k8s without proper error message:
```
28m         Warning   FailedPreStopHook   pod/ibm-spectrum-scale-csi-4lmb2                              PreStopHook failed
```
likely due to a hook can be called multiple times and hook fails if the container is already in a terminated/completed state.

**Comment by saurabhwani5** [comment]
Hii @amdabhad , this FailedPreStopHook Warning is coming in upgrade path (e.g. 2.8.0-> 2.9.0) and if you check the pod (oc get driver_pod -oyaml ) then in driver registrar ` /var/lib/kubelet/plugins/spectrumscale.csi.ibm.com/csi.sock` visible

**Comment by amdabhad** [comment]
The events like below from k8s without proper error message are likely due to a hook can be called multiple times and hook fails if the container is already in a terminated/completed state.:
```
28m         Warning   FailedPreStopHook   pod/ibm-spectrum-scale-csi-4lmb2                              PreStopHook failed
```
Checking why driver-registrar has the hook after upgrade.

**Comment by amdabhad** [comment]
The mergo package we use doesnot override the nil/empty struct value of spec.lifecycle in driver-registrar while upgrade - 
https://github.com/imdario/mergo/issues/185. So until this is fixed for upgrade to 2.9.0 case the preStop hook will be seen in driver-registrar too.

**Comment by saurabhwani5** [comment]
Not getting FailedPreStopHook event when driver pod is deleted on ocp and k8s 
```
[k8s ~]# oc get pods
NAME                                                  READY   STATUS    RESTARTS      AGE
ibm-spectrum-scale-csi-attacher-759c9d7bd4-9rntz      1/1     Running   1 (18h ago)   18h
ibm-spectrum-scale-csi-attacher-759c9d7bd4-fhbnd      1/1     Running   0             18h
ibm-spectrum-scale-csi-operator-5d5c8588bb-rdrfk      1/1     Running   0             18h
ibm-spectrum-scale-csi-provisioner-58b4d75cfd-t5rds   1/1     Running   1 (18h ago)   18h
ibm-spectrum-scale-csi-qtkj2                          3/3     Running   0             18h
ibm-spectrum-scale-csi-resizer-6577bb5ddd-qkwlk       1/1     Running   1 (18h ago)   18h
ibm-spectrum-scale-csi-snapshotter-85b4fcf68c-vxb4j   1/1     Running   1 (18h ago)   18h
ibm-spectrum-scale-csi-xb9w5                          3/3     Running   0             18h
[k8s ~]# oc get cso
NAME                     VERSION   SUCCESS
ibm-spectrum-scale-csi   2.10.0    True
[k8s ~]# oc describe pod | grep quay
    Image:         quay.io/ibm-spectrum-scale-dev/ibm-spectrum-scale-csi-operator:v2.10.0-220923
    Image ID:      quay.io/ibm-spectrum-scale-dev/ibm-spectrum-scale-csi-operator@sha256:52a24a97a9ad649603c8708d874c538fc22ef067767761fef9ffae091846695e
      CSI_DRIVER_IMAGE:      quay.io/ibm-spectrum-scale-dev/ibm-spectrum-scale-csi-driver:v2.10.0-220923
    Image:         quay.io/ibm-spectrum-scale-dev/ibm-spectrum-scale-csi-driver:v2.10.0-220923
    Image ID:      quay.io/ibm-spectrum-scale-dev/ibm-spectrum-scale-csi-driver@sha256:05837457d22e79593aa476f61374e640b9762e3e53b57b433d08140de70be96d
    Image:         quay.io/ibm-spectrum-scale-dev/ibm-spectrum-scale-csi-driver:v2.10.0-220923
    Image ID:      quay.io/ibm-spectrum-scale-dev/ibm-spectrum-scale-csi-driver@sha256:05837457d22e79593aa476f61374e640b9762e3e53b57b433d08140de70be96d
[k8s ~]# oc delete pod ibm-spectrum-scale-csi-xb9w5
pod "ibm-spectrum-scale-csi-xb9w5" deleted
[k8s ~]# oc get events | grep FailedPreStopHook
```
